{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a2fa4c7c-f257-4b84-a0e6-9757599ac996",
   "metadata": {},
   "source": [
    "# X-Y allele frequency dynamics under meiotic drive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48a2de4f-f549-4d86-ba5f-48a7f80e6259",
   "metadata": {
    "slideshow": {
     "slide_type": "skip"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from numpy.random import choice\n",
    "from random import random, randint, sample, seed\n",
    "from collections import defaultdict\n",
    "import scipy.stats\n",
    "import h5py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "233d6855-e5f4-495e-98d0-5ee2ec8ee69a",
   "metadata": {
    "slideshow": {
     "slide_type": "skip"
    }
   },
   "outputs": [],
   "source": [
    "# import multiprocessing\n",
    "import multiprocess\n",
    "import sys\n",
    "import collections\n",
    "from random import random\n",
    "from time import sleep\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from tqdm.notebook import trange, tqdm\n",
    "\n",
    "#import matplotlib.animation as animation\n",
    "\n",
    "from matplotlib import animation, rc\n",
    "from IPython.display import HTML\n",
    "\n",
    "from matplotlib.lines import Line2D\n",
    "\n",
    "import seaborn as sns\n",
    "sns.set()\n",
    "sns.set_style('ticks')\n",
    "# sns.set_style('dark')\n",
    "\n",
    "#colors = ['b', 'g', 'r', 'c', 'm', 'y', 'k', 'w']\n",
    "colors = sns.color_palette(\"Set1\", 9)\n",
    "#colors = sns.color_palette(\"rainbow\", 8)\n",
    "\n",
    "# scale down size of default plots\n",
    "sns.set_context(\"paper\")\n",
    "import matplotlib as mpl\n",
    "scale = 0.8\n",
    "d = dict([(k, v*scale) for (k, v) in sns.plotting_context('paper').items()])\n",
    "d['figure.figsize'] = [5.4, 3.5]\n",
    "mpl.rcParams.update(d)\n",
    "\n",
    "\n",
    "#sns.palplot(colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca9b67a8-76d3-4cec-92ee-12ad99cffe9e",
   "metadata": {
    "slideshow": {
     "slide_type": "skip"
    }
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "# scale down size of default plots\n",
    "sns.set_context(\"paper\")\n",
    "import matplotlib as mpl\n",
    "scale = 0.8\n",
    "d = dict([(k, v*scale) for (k, v) in sns.plotting_context('paper').items()])\n",
    "d['figure.figsize'] = [5.4, 3.5]\n",
    "mpl.rcParams.update(d)\n",
    "\n",
    "# Make inline plots vector graphics instead of raster graphics\n",
    "from matplotlib_inline.backend_inline import set_matplotlib_formats\n",
    "set_matplotlib_formats('retina', 'png')\n",
    "# set_matplotlib_formats('svg', 'pdf')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e14e0e70-fb59-4f17-a82a-a87fd9883ea4",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "We can sample drives between 0 and 1 from a beta:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "725e72ac-837e-4439-a5aa-2ee16a152955",
   "metadata": {
    "hide_input": false,
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "x = np.linspace(0.0001, 0.9999, 100)\n",
    "for a, b in [(5, 1), (2, 5), (1, 5), (1, 1)]:\n",
    "    plt.plot(x, scipy.stats.beta.pdf(x, a, b), label='alpha: {}, beta: {}'.format(a, b))\n",
    "plt.legend(loc='upper right') ;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f94069a-7faf-436b-a917-ad222a494fcf",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "... but it does not matter which distribution such allele specivic values are drawn from. When we compute drive as `x_drive / (x_drive + y_drive)` this is always uniform:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63272d5e-0816-40b8-9bce-4868d5bb5b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = np.linspace(0.0001, 0.9999, 100000)\n",
    "for a, b in [(0, 1), (10, 1), (5, 1)]:\n",
    "    x_drive = abs(np.random.normal(a, b, size=len(x)))\n",
    "    plt.hist(x_drive, density=True, bins=200, alpha=0.2, label='alpha: {}, beta: {}'.format(a, b))\n",
    "plt.legend() ;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbb873e2-09ce-41c9-83ca-15df8f1a093f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2)\n",
    "x = np.linspace(0.0001, 0.9999, 100000)\n",
    "for a, b in [(15, math.sqrt(15/2)), (10, math.sqrt(10)), (5, math.sqrt(5))]:\n",
    "    x_drive = abs(np.random.normal(a, b, size=len(x)))\n",
    "    y_drive = abs(np.random.normal(a, b, size=len(x)))\n",
    "    ax1.hist(x_drive, density=True, bins=200, alpha=0.2)\n",
    "    ax2.hist(x_drive / (x_drive + y_drive), density=True, bins=200, alpha=0.2, label='alpha: {}, beta: {}'.format(a, b))\n",
    "plt.legend() ;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1feac541-0d99-4bea-944c-30d6a28b598d",
   "metadata": {},
   "source": [
    "- This notebook simulates 10 (or any other) number of loci spread across the X chromosome. \n",
    "- Each locus interacts with a competing locus on the Y chromosome.\n",
    "- All loci start out with a \"drive\" values that determines its functional relationship to each allele at the competing locus. The initial drive value is sampled from a normal distribution with a mean of 100 and a standard deviation of 1.\n",
    "- The X chromosome recombines once in each female meiosis.\n",
    "- In each male meiosis, the meiotic drive (female bias) is computed as `x_allele_drive_value / (x_allele_drive_value + y_allele_drive_value)`. I.e. the relative ability fo the two alleles to drive against each other.\n",
    "- All X and Y alleles allele mutates at the same rate. This independently perturbs each all associated drive values by a value sampled form a normal distribution with a mean of 0 and a standard deviation of 20.\n",
    "- The motivation for indpendence of drive value pertubations is that with a large number of interacting genes outside the model each drive value (e.g. expression level) may have unique properties. I.e. the alterantive would be to only update the mutating allele and assume that this change completely captures its relation to the alleles on the competing locus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0039ccdd-e01d-47f0-9bcc-aa4baf6ea109",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(2, 2, figsize=(7, 3))\n",
    "x = np.linspace(0.0001, 0.9999, 100000)\n",
    "init_x_drives = np.random.normal(100, 1, size=len(x))\n",
    "init_y_drives = np.random.normal(100, 1, size=len(x))\n",
    "#ax1.hist(init_x_drives, density=True, bins=200, alpha=0.2)\n",
    "axes[0, 0].hist(init_x_drives, density=True, bins=100)\n",
    "axes[0, 0].set_title('Allele drives value after 1 mutation')\n",
    "axes[1, 0].hist(init_x_drives / (init_x_drives + init_y_drives), density=True, bins=100)\n",
    "axes[1, 0].set_title('Female distortions after 1 mutation')\n",
    "for i in range(99):\n",
    "    x_drive = np.random.normal(0, 20, size=len(x))\n",
    "    y_drive = np.random.normal(0, 20, size=len(x))\n",
    "    init_x_drives += x_drive\n",
    "    init_y_drives += y_drive\n",
    "    init_x_drives = abs(init_x_drives)\n",
    "    init_y_drives = abs(init_y_drives)\n",
    "axes[0, 1].hist(init_x_drives, density=True, bins=100)\n",
    "axes[0, 1].set_title('Allele drives value after 100 mutations')\n",
    "axes[1, 1].hist(init_x_drives / (init_x_drives + init_y_drives), density=True, bins=100)\n",
    "axes[1, 1].set_title('Female distortions after 100 mutations')\n",
    "plt.suptitle('Distribition of possible values assuming no contraint by selection')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26b0f54b-465e-468b-a2a1-90d6b16e61f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "class Node:\n",
    "    \n",
    "    def __init__(self, nodeid, label, generation, parent):\n",
    "        self.nodeid = nodeid\n",
    "        self.label = label\n",
    "        self.generation = generation\n",
    "        self.parent = parent\n",
    "        self.anc_child = None\n",
    "        self.der_child = None\n",
    "        \n",
    "    def __repr__(self):\n",
    "        p = self.parent.nodeid if self.parent is not None else None\n",
    "        a = self.anc_child.nodeid if self.anc_child is not None else None\n",
    "        d = self.der_child.nodeid if self.der_child is not None else None\n",
    "        return f\"ID: {self.nodeid}, Parent: {p}, Anc child: {a}, Der child: {d}, label: {self.label}, generation: {self.generation}\"\n",
    "\n",
    "\n",
    "#terminal_nodes = deque()\n",
    "terminal_nodes = dict()\n",
    "all_nodes = dict()\n",
    "\n",
    "# tmrca node joining both populations\n",
    "tmrca_node = Node(nodeid=0, label=None, generation=0, parent=None)\n",
    "all_nodes[tmrca_node.label] = tmrca_node\n",
    "\n",
    "# initialize\n",
    "node = Node(nodeid=len(all_nodes), label=0, generation=0, parent=tmrca_node)\n",
    "all_nodes[node.label] = node\n",
    "terminal_nodes[node.label] = node\n",
    "\n",
    "\n",
    "def update_tree(parent_label, child_label, generation, terminal_nodes, all_nodes):\n",
    "    \n",
    "    # mutate allele 1:\n",
    "    mutating_node = terminal_nodes[parent_label]\n",
    "\n",
    "    # append new nodes\n",
    "    anc_child_node = Node(nodeid=len(all_nodes), label=mutating_node.label, generation=generation, parent=mutating_node)\n",
    "    all_nodes[anc_child_node.label] = anc_child_node\n",
    "    terminal_nodes[anc_child_node.label] = anc_child_node\n",
    "    mutating_node.anc_child = anc_child_node\n",
    "    der_child_node =  Node(nodeid=len(all_nodes), label=child_label, generation=generation, parent=mutating_node)\n",
    "    all_nodes[der_child_node.label] = der_child_node\n",
    "    terminal_nodes[der_child_node.label] = der_child_node\n",
    "    mutating_node.der_child = der_child_node\n",
    "    \n",
    "update_tree(0, 777, 10, terminal_nodes, all_nodes)\n",
    "update_tree(777, 888, 5, terminal_nodes, all_nodes)\n",
    "update_tree(0, 999, 15, terminal_nodes, all_nodes)\n",
    "\n",
    "\n",
    "# for i in range(len(all_nodes)):\n",
    "#     print(all_nodes[i])\n",
    "# print()\n",
    "# for label in terminal_nodes:\n",
    "#     print(terminal_nodes[label])\n",
    "# print()\n",
    "\n",
    "node1 = terminal_nodes[888]\n",
    "node2 = terminal_nodes[999]\n",
    "\n",
    "def parents(node):\n",
    "    if node.parent is None:\n",
    "        return []\n",
    "    return [node.label] + parents(node.parent)\n",
    "\n",
    "def available_coal_span(node1, node2):\n",
    "\n",
    "    a = parents(node1)\n",
    "    b = parents(node2)\n",
    "    for i in range(min(len(a), len(b))):\n",
    "        if a[i] != b[i]:\n",
    "            max_coal_gen = all_nodes[a[i-1]].generation\n",
    "        min_coal_gen = min(all_nodes[a[i]].generation, all_nodes[b[i]].generation)\n",
    "    return max_coal_gen, min_coal_gen\n",
    "\n",
    "def expected_coal_time(node1, node2, demography, freq_trajectory):\n",
    "    \n",
    "    if node1.label == node2.label:\n",
    "        max_gen, min_gen = node1.generation, len(demography)\n",
    "    else:\n",
    "        max_gen, min_gen = available_coal_span(node1, node2)\n",
    "    exptime = 0\n",
    "    prev_exptime = 1\n",
    "    for gen in range(max_gen, min_gen - max_gen):\n",
    "        exptime, prev_exptime = exptime + 1 + ((1-1/(demography[gen]*freq_trajectory[gen])) * prev_exptime), exptime\n",
    "    return max_gen + exptime\n",
    "    \n",
    "# pop1_size = 100\n",
    "# pop1_demog = [[pop1_size, 10000]]\n",
    "# pop1_size_history = [x for (size, gens) in pop1_demog for x in [size]*gens]\n",
    "\n",
    "# freqs = dict(zip(x_freqs['pop1'][0][0], y_freqs['pop1'][0][0]))\n",
    "\n",
    "# pwdists = defaultdict(dict)\n",
    "# for a, b in zip(all_nodes.keys(), all_nodes.keys()):\n",
    "#     n1 = all_nodes[a]\n",
    "#     n2 = all_nodes[b]\n",
    "#     pwdists[a][b] = expected_coal_time(n1, n2, pop1_size_history, freqs)\n",
    "\n",
    "# # lookup table for computing mean pwdist of all indivs:\n",
    "# print(pwdists)    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0de4352c-815c-446f-b444-92d0830383dd",
   "metadata": {},
   "source": [
    "## TODO:\n",
    "\n",
    "- The cumulative changes becomes relatively smaller the further from 0 a drive is (as long time pases, changes will become very small)\n",
    "- Add autosomes\n",
    "- Track selection coeficients\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec127302-f1e1-48c2-ace2-0889be5e4de2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## seed(7)\n",
    "\n",
    "class Allele:\n",
    "    __slots__ = ('chrom', 'label', 'val')\n",
    "    \n",
    "    def __init__(self, chrom, label, val):\n",
    "        self.chrom = chrom\n",
    "        self.label = label\n",
    "        self.val = val\n",
    "    \n",
    "class Male:\n",
    "    __slots__ = ('x_alleles', 'y_alleles', 'parent_female_offspring_ratio')\n",
    "        \n",
    "    def __init__(self, x_alleles, y_alleles, parent_female_offspring_ratio):\n",
    "        # assert x_allele.chrom == 'X'\n",
    "        # assert y_allele.chrom == 'Y'\n",
    "        self.x_alleles = x_alleles\n",
    "        self.y_alleles = y_alleles\n",
    "        self.parent_female_offspring_ratio = parent_female_offspring_ratio        \n",
    "\n",
    "    def __add__(self, female):\n",
    "      \n",
    "        # male gametes\n",
    "        x_proportion = 0\n",
    "        for locus in range(LOCI):\n",
    "            x_drive = DRIVES[locus][self.x_alleles[locus].label][self.y_alleles[locus].label]\n",
    "            y_drive = DRIVES[locus][self.y_alleles[locus].label][self.x_alleles[locus].label]\n",
    "            x_proportion += x_drive / (x_drive + y_drive)\n",
    "        x_proportion /= LOCI\n",
    "        \n",
    "        if random() < x_proportion:\n",
    "            sperm_alleles = self.x_alleles[:]\n",
    "        else:\n",
    "            sperm_alleles = self.y_alleles[:]\n",
    "        \n",
    "        # female gametes\n",
    "        lst = []\n",
    "        recombination_point = random() * len(female.x1_alleles)\n",
    "        for i, (a, b) in enumerate(zip(female.x1_alleles, female.x2_alleles)):\n",
    "            if i < recombination_point:\n",
    "                lst.append([a, b])\n",
    "            else:\n",
    "                lst.append([b, a])\n",
    "            # if random() < 0.5:\n",
    "            #     lst.append([a, b])\n",
    "            # else:\n",
    "            #     lst.append([b, a])\n",
    "        x1_alleles, x2_alleles = map(list, zip(*lst))\n",
    "            \n",
    "        if random() < 0.5:\n",
    "            egg_alleles = x1_alleles\n",
    "        else:        \n",
    "            egg_alleles = x2_alleles\n",
    "  \n",
    "        # make new individual\n",
    "        if sperm_alleles[0].chrom == 'Y':\n",
    "            return Male(egg_alleles, sperm_alleles, x_proportion)\n",
    "        else:\n",
    "            return Female(egg_alleles, sperm_alleles, x_proportion)\n",
    "            \n",
    "class Female:\n",
    "    __slots__ = ('x1_alleles', 'x2_alleles', 'parent_female_offspring_ratio')\n",
    "\n",
    "    def __init__(self, x1_alleles, x2_alleles, parent_female_offspring_ratio):\n",
    "        # assert x1_allele.chrom == 'X'\n",
    "        # assert x2_allele.chrom == 'X'\n",
    "        self.x1_alleles = x1_alleles\n",
    "        self.x2_alleles = x2_alleles\n",
    "        self.parent_female_offspring_ratio = parent_female_offspring_ratio\n",
    "\n",
    "    def __add__(self, male):\n",
    "        return male + self\n",
    "    \n",
    "\n",
    "class Population:\n",
    "    __slots__ = ('males', 'females', 'alpha', 'beta', 'pop_size', 'drive_coef', 'allele_val_map', 'mutation_tree', 'generation')\n",
    "        \n",
    "    def __init__(self, males=[], females=[], alpha=5, beta=1, generation=1):\n",
    "        self.generation = generation\n",
    "        self.males = males\n",
    "        self.females = females\n",
    "        self.alpha = alpha\n",
    "        self.beta = beta\n",
    "        self.pop_size = len(self.males) + len(self.females)\n",
    "        self.drive_coef = defaultdict(dict)\n",
    "        self.allele_val_map = defaultdict(dict)\n",
    "        self.mutation_tree = dict()\n",
    "        \n",
    "    def initialize(self, pop_size, val=None):\n",
    "        assert not pop_size % 2, 'initial pop_size must be even'\n",
    "        self.males = [Male([Allele('X', CUR_XLABEL[locus], val) for locus in range(LOCI)], \n",
    "                           [Allele('Y', CUR_YLABEL[locus], val) for locus in range(LOCI)], np.nan) for _ in range(pop_size//2)]\n",
    "        self.females = [Female([Allele('X', CUR_XLABEL[locus], val) for locus in range(LOCI)], \n",
    "                               [Allele('X', CUR_XLABEL[locus], val) for locus in range(LOCI)], np.nan) for _ in range(pop_size//2)]\n",
    "\n",
    "        for locus in range(LOCI):\n",
    "            self.update_drives(locus, None, CUR_XLABEL[locus])\n",
    "            self.update_drives(locus, None, CUR_YLABEL[locus])\n",
    "            self.allele_val_map[locus][CUR_XLABEL[locus]] = val\n",
    "            self.allele_val_map[locus][CUR_YLABEL[locus]] = val\n",
    "            self.update_tree(locus, -2, CUR_XLABEL[locus], self.generation) # root of x accross pops\n",
    "            self.update_tree(locus, -1, CUR_YLABEL[locus], self.generation) # root of x accross pops\n",
    "\n",
    "#     def update_drives(self, label):\n",
    "#         # assert label >= len(self.DRIVES)\n",
    "\n",
    "#         nrow = len(self.DRIVES[0]) if self.DRIVES else 1\n",
    "#         row = np.random.normal(self.alpha, self.beta, size=nrow).tolist()\n",
    "#         self.DRIVES.append(row)\n",
    "        \n",
    "#         ncol = len(self.DRIVES)\n",
    "#         col = np.random.normal(self.alpha, self.beta, size=ncol).tolist()\n",
    "#         for row, val in zip(self.DRIVES, col):\n",
    "#             row.append(val)\n",
    "            \n",
    "    def update_drives(self, locus, parent_label, label):  \n",
    "        global DRIVES\n",
    "        \n",
    "        assert label not in DRIVES[locus].columns.values, (label, DRIVES[locus].columns.values)\n",
    "\n",
    "        cumulative = True\n",
    "        \n",
    "        nrows, ncols = DRIVES[locus].shape   \n",
    "        if cumulative:\n",
    "            if parent_label is None:\n",
    "                parent2other_drives = np.random.normal(self.alpha, self.beta, size=nrows)\n",
    "            else:\n",
    "                parent2other_drives = DRIVES[locus][parent_label]\n",
    "            DRIVES[locus][label] = np.abs(parent2other_drives + np.random.normal(0, 20, size=nrows))\n",
    "        else:\n",
    "            DRIVES[locus][label] = np.random.normal(self.alpha, self.beta, size=nrows)\n",
    "\n",
    "        nrows, ncols = DRIVES[locus].shape \n",
    "        if cumulative:        \n",
    "            if parent_label is None:\n",
    "                other2parent_drives = np.random.normal(self.alpha, self.beta, size=ncols)\n",
    "            else:\n",
    "                other2parent_drives = DRIVES[locus].loc[parent_label]\n",
    "            new_row = pd.DataFrame.from_records([np.abs(other2parent_drives + np.random.normal(0, 20, size=ncols))], \n",
    "                                                columns=DRIVES[locus].columns)                \n",
    "        else:\n",
    "            new_row = pd.DataFrame.from_records([np.random.normal(self.alpha, self.beta, size=ncols)], \n",
    "                                                columns=DRIVES[locus].columns)\n",
    "        new_row.index = [label]\n",
    "        DRIVES[locus] = pd.concat([DRIVES[locus], new_row])\n",
    "        DRIVES[locus] = DRIVES[locus].copy() \n",
    "        \n",
    "    def update_tree(self, locus, parent_label, child_label, generation):\n",
    "    \n",
    "        global NODES\n",
    "\n",
    "        # mutate allele 1:\n",
    "        mutating_node = NODES[locus][parent_label]\n",
    "\n",
    "        # Only need NODES, but only add the der child node not the anc node\n",
    "        \n",
    "        # append new nodes\n",
    "        # anc_child_node = Node(nodeid=len(NODES[locus]), label=mutating_node.label, generation=generation, parent=mutating_node)\n",
    "        # NODES[locus][anc_child_node.label] = anc_child_node\n",
    "        # mutating_node.anc_child = anc_child_node\n",
    "        der_child_node =  Node(nodeid=len(NODES[locus]), label=child_label, generation=generation, parent=mutating_node)\n",
    "        NODES[locus][der_child_node.label] = der_child_node\n",
    "        mutating_node.der_child = der_child_node\n",
    "                \n",
    "    def resample(self, new_pop_size=None):\n",
    "        if new_pop_size is None:\n",
    "            new_pop_size = self.pop_size\n",
    "        new_males = list()\n",
    "        new_females = list()\n",
    "        \n",
    "        for i in range(new_pop_size):\n",
    "            male = self.males[int(random() * len(self.males))]\n",
    "            female = self.females[int(random() * len(self.females))]\n",
    "            child = male + female\n",
    "                        \n",
    "            if isinstance(child, Male):\n",
    "                new_males.append(child)\n",
    "            else:\n",
    "                new_females.append(child)       \n",
    "                \n",
    "        return Population(new_males, new_females, generation=self.generation+1)\n",
    "\n",
    "    def mutate(self, x_mut_prob, y_mut_prob):\n",
    "\n",
    "        global CUR_XLABEL\n",
    "        global CUR_YLABEL\n",
    "\n",
    "        for i in range(len(self.males)):\n",
    "            for locus in range(LOCI):\n",
    "                if random() < x_mut_prob:\n",
    "                    CUR_XLABEL[locus] += 2\n",
    "                    val = self.males[i].x_alleles[locus].val\n",
    "\n",
    "                    parent_label = self.males[i].x_alleles[locus].label\n",
    "                    child_label = CUR_XLABEL[locus]\n",
    "                    self.males[i].x_alleles[locus] = Allele('X', child_label, val)\n",
    "                    self.update_drives(locus, parent_label, child_label)\n",
    "                    self.allele_val_map[locus][child_label] = val\n",
    "                    self.update_tree(locus, parent_label, child_label, self.generation)\n",
    "                    \n",
    "                    # self.males[i].x_alleles[locus] = Allele('X', CUR_XLABEL[locus], val)\n",
    "                    # self.update_drives(locus, CUR_XLABEL[locus])\n",
    "                    # self.allele_val_map[locus][CUR_XLABEL[locus]] = val\n",
    "                    # self.update_tree(locus, CUR_XLABEL[locus]-2, CUR_XLABEL[locus], self.generation)\n",
    "\n",
    "                if random() < y_mut_prob:\n",
    "                    CUR_YLABEL[locus] += 2\n",
    "                    val = self.males[i].y_alleles[locus].val\n",
    "                    \n",
    "                    parent_label = self.males[i].y_alleles[locus].label\n",
    "                    child_label = CUR_YLABEL[locus]\n",
    "                    self.males[i].y_alleles[locus] = Allele('Y', child_label, val)\n",
    "                    self.update_drives(locus, parent_label, child_label)\n",
    "                    self.allele_val_map[locus][child_label] = val\n",
    "                    self.update_tree(locus, parent_label, child_label, self.generation)               \n",
    "                    # self.males[i].y_alleles[locus] = Allele('Y', CUR_YLABEL[locus], val)\n",
    "                    # self.update_drives(locus, CUR_YLABEL[locus])\n",
    "                    # self.allele_val_map[locus][CUR_YLABEL[locus]] = val\n",
    "                    # self.update_tree(locus, CUR_YLABEL[locus]-2, CUR_YLABEL[locus], self.generation)\n",
    "                    \n",
    "        for i in range(len(self.females)):\n",
    "            for locus in range(LOCI):\n",
    "                if random() < x_mut_prob:\n",
    "                    CUR_XLABEL[locus] += 2\n",
    "                    val = self.females[i].x1_alleles[locus].val\n",
    "                    \n",
    "                    parent_label = self.females[i].x1_alleles[locus].label\n",
    "                    child_label = CUR_XLABEL[locus]\n",
    "                    self.females[i].x1_alleles[locus] = Allele('X', child_label, val)\n",
    "                    self.update_drives(locus, parent_label, child_label)\n",
    "                    self.allele_val_map[locus][child_label] = val\n",
    "                    self.update_tree(locus, parent_label, child_label, self.generation)                    \n",
    "                    # self.females[i].x1_alleles[locus] = Allele('X', CUR_XLABEL[locus], val)\n",
    "                    # self.update_drives(locus, CUR_XLABEL[locus])\n",
    "                    # self.allele_val_map[locus][CUR_XLABEL[locus]] = val\n",
    "                    # self.update_tree(locus, CUR_XLABEL[locus]-2, CUR_XLABEL[locus], self.generation)                    \n",
    "\n",
    "                if random() < x_mut_prob:\n",
    "                    CUR_XLABEL[locus] += 2\n",
    "                    val = self.females[i].x2_alleles[locus].val\n",
    "                    \n",
    "                    parent_label = self.females[i].x2_alleles[locus].label\n",
    "                    child_label = CUR_XLABEL[locus]\n",
    "                    self.females[i].x2_alleles[locus] = Allele('X', child_label, val)\n",
    "                    self.update_drives(locus, parent_label, child_label)\n",
    "                    self.allele_val_map[locus][child_label] = val\n",
    "                    self.update_tree(locus, parent_label, child_label, self.generation) \n",
    "                    # self.females[i].x2_alleles[locus] = Allele('X', CUR_XLABEL[locus], val)\n",
    "                    # self.update_drives(locus, CUR_XLABEL[locus])\n",
    "                    # self.allele_val_map[locus][CUR_XLABEL[locus]] = val\n",
    "                    # self.update_tree(locus, CUR_XLABEL[locus]-2, CUR_XLABEL[locus], self.generation)                    \n",
    "                    \n",
    "\n",
    "    def live_alleles(self, locus):\n",
    "        alleles = [male.x_alleles[locus].label for male in self.males] + \\\n",
    "            [male.y_alleles[locus].label for male in self.males] + \\\n",
    "            [female.x1_alleles[locus].label for female in self.females] + \\\n",
    "            [female.x2_alleles[locus].label for female in self.females] \n",
    "        return sorted(set(alleles))\n",
    "                    \n",
    "    def x_freqs(self, locus):\n",
    "        alleles = [male.x_alleles[locus].label for male in self.males] + \\\n",
    "            [female.x1_alleles[locus].label for female in self.females] + \\\n",
    "            [female.x2_alleles[locus].label for female in self.females]\n",
    "        unique_elements, counts_elements = np.unique(np.array(alleles), return_counts=True)\n",
    "        return unique_elements, counts_elements / counts_elements.sum()\n",
    "\n",
    "    def y_freqs(self, locus):\n",
    "        alleles = [male.y_alleles[locus].label for male in self.males]\n",
    "        unique_elements, counts_elements = np.unique(np.array(alleles), return_counts=True)\n",
    "        return unique_elements, counts_elements / counts_elements.sum()\n",
    "\n",
    "    def male_freq(self):\n",
    "        return len(self.males) / (len(self.males) + len(self.females)) \n",
    "    \n",
    "    def sex_ratio_std(self):\n",
    "        return np.array([indiv.parent_female_offspring_ratio for indiv in self.males + self.females]).std()\n",
    "    \n",
    "\n",
    "class Node:\n",
    "    \n",
    "    def __init__(self, nodeid, label, generation, parent):\n",
    "        self.nodeid = nodeid\n",
    "        self.label = label\n",
    "        self.generation = generation\n",
    "        self.parent = parent\n",
    "        self.anc_child = None\n",
    "        self.der_child = None\n",
    "        \n",
    "    def __repr__(self):\n",
    "        p = self.parent.nodeid if self.parent is not None else None\n",
    "        a = self.anc_child.nodeid if self.anc_child is not None else None\n",
    "        d = self.der_child.nodeid if self.der_child is not None else None\n",
    "        return f\"ID: {self.nodeid}, Parent: {p}, Anc child: {a}, Der child: {d}, label: {self.label}, generation: {self.generation}\"\n",
    "\n",
    "\n",
    "###########################\n",
    "\n",
    "# # bottleneck\n",
    "# pop1_demog = [[10000, 100], [1000, 100], [10000, 100]]\n",
    "\n",
    "pop1_size = 100\n",
    "pop2_size = 100\n",
    "pop1_demog = [[pop1_size, 10000]]\n",
    "pop2_demog = [[pop2_size, 10000]]\n",
    "x_mutation_rate = 0.001\n",
    "y_mutation_rate = 0.001\n",
    "generations = 200\n",
    "introgression_generations = []\n",
    "introgressed_males = 0\n",
    "introgressed_females = 0\n",
    "\n",
    "# #################################################\n",
    "# # THIS WORKS WELL (stored as test0.h5 though test9.h5)\n",
    "# pop1_size = 10000\n",
    "# pop2_size = 10000\n",
    "# pop1_demog = [[pop1_size, 10000]]\n",
    "# pop2_demog = [[pop2_size, 10000]]\n",
    "# x_mutation_rate = 0.00001\n",
    "# y_mutation_rate = 0.00001\n",
    "# generations = 2000\n",
    "# introgression_generations = [1500]\n",
    "# introgressed_males = 100\n",
    "# introgressed_females = 100\n",
    "# #################################################\n",
    "\n",
    "# scaling = 0.01\n",
    "\n",
    "# pop1_size = int(10000 * scaling)\n",
    "# pop2_size = int(10000 * scaling)\n",
    "# pop1_demog = [[pop1_size, 10000]]\n",
    "# pop2_demog = [[pop2_size, 10000]]\n",
    "# x_mutation_rate = 0.001 / scaling\n",
    "# y_mutation_rate = 0.001 / scaling\n",
    "# generations = int(3000 / scaling)\n",
    "# introgression_generation = int(2000 / scaling)\n",
    "# introgressed_males = int(500 / scaling)\n",
    "# introgressed_females = int(500 / scaling)\n",
    "\n",
    "# pop1_demog = [[100, 10000]]\n",
    "# pop2_demog = [[100, 10000]]\n",
    "# x_mutation_rate = 0.01\n",
    "# y_mutation_rate = 0.01\n",
    "# generations = 300\n",
    "# introgression_generation = 200\n",
    "# introgressed_males = 5\n",
    "# introgressed_females = 5\n",
    "\n",
    "nr_loci = 10\n",
    "# alpha = 5\n",
    "# beta = 1\n",
    "alpha = 100\n",
    "beta = 1\n",
    "\n",
    "for run in range(1):\n",
    "\n",
    "    hdf5_file_name = f'devel{run}.h5'\n",
    "\n",
    "    ###########################\n",
    "\n",
    "    LOCI = nr_loci\n",
    "\n",
    "    DRIVES = [pd.DataFrame() for _ in range(LOCI)]\n",
    "\n",
    "    NODES = defaultdict(dict)\n",
    "    for locus in range(LOCI):\n",
    "        # node joining x in both populations\n",
    "        ancestral_node = Node(nodeid=0, label=-2, generation=0, parent=None) # Problem: gen=0 implies instant coal of alleles from two pops at time 0\n",
    "        all_nodes[tmrca_node.label] = tmrca_node\n",
    "        NODES[locus][ancestral_node.label] = ancestral_node\n",
    "        # node joining y in both populations\n",
    "        ancestral_node = Node(nodeid=1, label=-1, generation=0, parent=None) # Problem: gen=0 implies instant coal of alleles from two pops at time 0\n",
    "        all_nodes[tmrca_node.label] = tmrca_node\n",
    "        NODES[locus][ancestral_node.label] = ancestral_node\n",
    "\n",
    "    CUR_XLABEL=[0 for _ in range(LOCI)]\n",
    "    CUR_YLABEL=[1 for _ in range(LOCI)]\n",
    "\n",
    "    pop1_size_history = [x for (size, gens) in pop1_demog for x in [size]*gens]\n",
    "    population1 = Population(alpha=alpha, beta=beta)\n",
    "    population1.initialize(pop1_size_history[0], val=1)\n",
    "\n",
    "    CUR_XLABEL=[2 for _ in range(LOCI)]\n",
    "    CUR_YLABEL=[3 for _ in range(LOCI)]\n",
    "\n",
    "    pop2_size_history = [x for (size, gens) in pop2_demog for x in [size]*gens]\n",
    "    population2 = Population(alpha=alpha, beta=beta)\n",
    "    population2.initialize(pop2_size_history.pop(0), val=2)\n",
    "\n",
    "    data_pop1_males = []\n",
    "    data_pop2_males = []\n",
    "    sex_ratio_std_pop1 = []\n",
    "    sex_ratio_std_pop2 = []\n",
    "    allele_val_map = defaultdict(dict)\n",
    "\n",
    "    for locus in range(LOCI):\n",
    "        allele_val_map[locus].update(population1.allele_val_map[locus])\n",
    "        allele_val_map[locus].update(population2.allele_val_map[locus])\n",
    "\n",
    "    with h5py.File(hdf5_file_name, 'w') as f:\n",
    "        # assert introgressed_males + introgressed_females < pop_size1 and introgressed_males + introgressed_females < pop_size2\n",
    "\n",
    "        f.attrs['loci'] = LOCI\n",
    "        f.attrs['alpha'] = alpha\n",
    "        f.attrs['beta'] = beta\n",
    "        f.attrs['x_mutation_rate'] = x_mutation_rate\n",
    "        f.attrs['y_mutation_rate'] = y_mutation_rate\n",
    "        f.attrs['generations'] = generations\n",
    "        pop1_gr = f.create_group('pop1')\n",
    "        pop1_gr.create_dataset('demography', data=np.array(pop1_size_history, dtype=np.int32), compression=9)    \n",
    "        pop1_samples_gr = pop1_gr.create_group('samples')\n",
    "        # pop1_gr.attrs['introgression_generations'] = introgression_generation\n",
    "        pop1_gr.create_dataset('introgression_generations', data=np.array(introgression_generations, dtype=np.int32), compression=9)    \n",
    "        # pop1_gr.attrs['introgression_generations'] = introgression_generation\n",
    "        pop1_gr.attrs['introgressed_males'] = introgressed_males\n",
    "        pop1_gr.attrs['introgressed_females'] = introgressed_females\n",
    "        pop2_gr = f.create_group('pop2')\n",
    "        pop2_gr.create_dataset('demography', data=np.array(pop2_size_history, dtype=np.int32), compression=9)    \n",
    "        pop2_samples_gr = pop2_gr.create_group('samples')\n",
    "        # pop2_gr.attrs['introgression_generations'] = introgression_generation\n",
    "        pop2_gr.create_dataset('introgression_generations', data=np.array(introgression_generations, dtype=np.int32), compression=9)    \n",
    "        pop2_gr.attrs['introgressed_males'] = introgressed_males\n",
    "        pop2_gr.attrs['introgressed_females'] = introgressed_females   \n",
    "\n",
    "        for gen in trange(1, generations+1):\n",
    "        # for gen in trange(generations):\n",
    "\n",
    "            sex_ratio_std_pop1.append(population1.sex_ratio_std())\n",
    "            sex_ratio_std_pop2.append(population2.sex_ratio_std())\n",
    "\n",
    "            if gen in introgression_generations:\n",
    "\n",
    "                # for distinguishing migrant alles\n",
    "                pop1_gr.create_dataset('alleles_at_introgression', data=np.array(population1.live_alleles(locus), dtype=np.int32), compression=9)\n",
    "                pop2_gr.create_dataset('alleles_at_introgression', data=np.array(population2.live_alleles(locus), dtype=np.int32), compression=9)\n",
    "\n",
    "                # male introgression\n",
    "                i = introgressed_males \n",
    "                population1.males, population2.males = \\\n",
    "                    population2.males[:i] + population1.males[i:], population1.males[:i] + population2.males[i:]\n",
    "                # female introgression\n",
    "                i = introgressed_females \n",
    "                population1.females, population2.females = \\\n",
    "                    population2.females[:i] + population1.females[i:], population1.females[:i] + population2.females[i:]        \n",
    "\n",
    "            for locus in range(LOCI):\n",
    "\n",
    "                if not gen % 50:\n",
    "                    live_labels_pop1 = population1.live_alleles(locus)\n",
    "                    live_labels_pop2 = population2.live_alleles(locus)\n",
    "                    live_labels = list(set(live_labels_pop1 + live_labels_pop2))\n",
    "                    DRIVES[locus] = DRIVES[locus].loc[live_labels, live_labels]     \n",
    "\n",
    "            chromosomes = [[a.label for a in m.y_alleles] for m in population1.males]\n",
    "            chromosomes.extend([[a.label for a in m.x_alleles] for m in population1.males])\n",
    "            chromosomes.extend([[a.label for a in f.x1_alleles] for f in population1.females])\n",
    "            chromosomes.extend([[a.label for a in f.x2_alleles] for f in population1.females])\n",
    "            pop1_samples_gr.create_dataset(f'{gen:0>5}', data=np.array(chromosomes), compression=9)\n",
    "            data_pop1_males.append(len(population1.males))\n",
    "\n",
    "            chromosomes = [[a.label for a in m.y_alleles] for m in population2.males]\n",
    "            chromosomes.extend([[a.label for a in m.x_alleles] for m in population2.males])\n",
    "            chromosomes.extend([[a.label for a in f.x1_alleles] for f in population2.females])\n",
    "            chromosomes.extend([[a.label for a in f.x2_alleles] for f in population2.females])\n",
    "            pop2_samples_gr.create_dataset(f'{gen:0>5}', data=np.array(chromosomes), compression=9)\n",
    "            data_pop2_males.append(len(population2.males))\n",
    "\n",
    "\n",
    "            pop_size = pop1_size_history[gen]\n",
    "            population1 = population1.resample(pop_size)\n",
    "            population1.mutate(x_mut_prob=x_mutation_rate, y_mut_prob=y_mutation_rate)\n",
    "\n",
    "            pop_size = pop2_size_history[gen]\n",
    "            population2 = population2.resample(pop_size)\n",
    "            population2.mutate(x_mut_prob=x_mutation_rate, y_mut_prob=y_mutation_rate)\n",
    "\n",
    "            for locus in range(LOCI):\n",
    "                allele_val_map[locus].update(population1.allele_val_map[locus])\n",
    "                allele_val_map[locus].update(population2.allele_val_map[locus])\n",
    "\n",
    "\n",
    "        pop1_gr.create_dataset('males', data=np.array(data_pop1_males, dtype=np.int32), compression=9)\n",
    "        pop2_gr.create_dataset('males', data=np.array(data_pop2_males, dtype=np.int32), compression=9)\n",
    "\n",
    "        pop1_gr.create_dataset('sex_ratio_std', data=np.array(sex_ratio_std_pop1), compression=9)\n",
    "        pop2_gr.create_dataset('sex_ratio_std', data=np.array(sex_ratio_std_pop2), compression=9)\n",
    "\n",
    "        maps_gr = f.create_group('allele_val_maps')\n",
    "        for locus in range(LOCI):\n",
    "            maps_gr.create_dataset(str(locus), data=np.array(list(allele_val_map[locus].items()), dtype=np.int32), compression=9)    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39728243-2952-40eb-9670-083feea6fcc9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "    \n",
    "def parents(node):\n",
    "    if node.parent is None:\n",
    "        return []\n",
    "    return [node.label] + parents(node.parent)\n",
    "\n",
    "def available_coal_span(locus, node1, node2):\n",
    "\n",
    "    a = parents(node1)\n",
    "    b = parents(node2)\n",
    "    assert len(a) > 1 and len(b) > 1, (a, b)\n",
    "    for i in range(min(len(a), len(b))):\n",
    "        if a[i] != b[i]:\n",
    "            max_coal_gen = NODES[locus][a[i-1]].generation\n",
    "            min_coal_gen = min(NODES[locus][a[i]].generation, NODES[locus][b[i]].generation)\n",
    "            return a[i-1], max_coal_gen, min_coal_gen\n",
    "\n",
    "def expected_coal_time(max_gen, min_gen, demography, freq_trajectory):\n",
    "\n",
    "    exptime = 0\n",
    "    prev_exptime = 1\n",
    "    for gen in range(max_gen, min_gen - max_gen):\n",
    "        try:\n",
    "            exptime, prev_exptime = exptime + 1 + ((1-1/(demography[gen]*freq_trajectory[gen])) * prev_exptime), exptime\n",
    "        except:\n",
    "            k = sorted(freq_trajectory.keys())\n",
    "            print(gen, max_gen, min_gen)\n",
    "            print(k, [freq_trajectory[x] for x in k])\n",
    "            raise\n",
    "    return max_gen + exptime\n",
    "\n",
    "# hdf5_file_name = f'test2.h5'\n",
    "\n",
    "with h5py.File(hdf5_file_name, 'r') as f: \n",
    "    \n",
    "    generations = f.attrs['generations']\n",
    "    loci = f.attrs['loci']\n",
    "\n",
    "    allele_val_maps = dict()\n",
    "    for locus in range(LOCI):\n",
    "        allele_val_maps[locus] = dict(f['allele_val_maps'][str(locus)][:])\n",
    "\n",
    "    np.array(list(allele_val_map[locus].items()), dtype=np.int32)\n",
    "    \n",
    "    x_freqs = dict((p, dict((l, defaultdict(list)) for l in range(loci))) for p in ['pop1', 'pop2'])\n",
    "    y_freqs = dict((p, dict((l, defaultdict(list)) for l in range(loci))) for p in ['pop1', 'pop2'])\n",
    "    male_ratio = dict((x, []) for x in ['pop1', 'pop2'])\n",
    "    nr_x_alleles = dict((l, defaultdict(list)) for l in ['pop1', 'pop2'])\n",
    "    nr_y_alleles = dict((x, []) for x in ['pop1', 'pop2'])\n",
    "    \n",
    "    sex_ratio_std = dict()\n",
    "    \n",
    "    for pop in tqdm(['pop1', 'pop2']):\n",
    "        \n",
    "        sex_ratio_std[pop] = f[pop]['sex_ratio_std'][:]\n",
    "        \n",
    "        nr_males = f[pop]['males'][:]\n",
    "        \n",
    "        for gen, (name, population) in tqdm(enumerate(f[pop]['samples'].items()), leave=False, total=generations):\n",
    "\n",
    "            male_ratio[pop].append(2 * nr_males[gen] / len(population))\n",
    "\n",
    "            nr_females = len(population) - nr_males[gen]\n",
    "            male_loci = population[:nr_males[gen]].T\n",
    "            female_loci = population[nr_males[gen]:].T\n",
    "\n",
    "            for i, locus in enumerate(male_loci):\n",
    "                alleles, counts = np.unique(locus, return_counts=True)\n",
    "                for a, fr in zip(alleles, counts/nr_males[gen]):\n",
    "                    x_freqs[pop][i][a].append(gen)\n",
    "                    y_freqs[pop][i][a].append(fr)\n",
    "            uniq, counts = np.unique(male_loci, axis=1, return_counts=True)\n",
    "            nr_y_alleles[pop].append(len(counts))\n",
    "\n",
    "            for i, locus in enumerate(female_loci):\n",
    "                alleles, counts = np.unique(locus, return_counts=True)\n",
    "                for a, fr in zip(alleles, counts/nr_females):\n",
    "                    x_freqs[pop][i][a].append(gen)\n",
    "                    y_freqs[pop][i][a].append(fr)\n",
    "                nr_x_alleles[pop][i].append(len(alleles))\n",
    "                \n",
    "#         pwdists = dict((i, defaultdict(dict)) for i in range(LOCI))\n",
    "#         if pop == 'pop1':\n",
    "#             demography = pop1_size_history\n",
    "#         else:\n",
    "#             demography = pop1_size_history\n",
    "#         for locus in range(LOCI):\n",
    "#             for a in NODES[locus].keys():\n",
    "#                 if a not in x_freqs[pop]:\n",
    "#                     continue\n",
    "#                 if len(x_freqs[pop][locus][a]) == 0:\n",
    "#                     continue\n",
    "#                 if x_freqs[pop][locus][a][-1] != generations-1: # TODO: -1 problem\n",
    "#                     continue                    \n",
    "#                 for b in NODES[locus].keys():\n",
    "#                     if b not in x_freqs[pop]:\n",
    "#                         continue\n",
    "\n",
    "#                     if len(x_freqs[pop][locus][b]) == 0:\n",
    "#                         continue\n",
    "#                     if x_freqs[pop][locus][b][-1] != generations-1: # TODO: -1 problem\n",
    "#                         continue\n",
    "#                     if a < 0 or b < 0:\n",
    "#                         continue\n",
    "                        \n",
    "#                     if bool(a % 2) != bool(b % 2): # do not compare x and y alleles\n",
    "#                         continue\n",
    "                        \n",
    "#                     n1 = NODES[locus][a]\n",
    "#                     n2 = NODES[locus][b]\n",
    "\n",
    "#                     if n1.label == n2.label:\n",
    "#                         anc_allele, max_gen, min_gen = n1.label, n1.generation, generations\n",
    "#                     else:\n",
    "#                         anc_allele, max_gen, min_gen = available_coal_span(locus, n1, n2)\n",
    "\n",
    "#                     # we need to somehow solve that the frequency trajectory needs to be spliced at the admixture point of alleles are from different pops originally\n",
    "#                     other_pop = 'pop2' if pop == 'pop1' else 'pop1'\n",
    "#                     # if anc_allele in x_freqs[other_pop] and min_gen >= introgression_generation:\n",
    "#                     print(max_gen, min_gen, introgression_generation)\n",
    "#                     if x_freqs[other_pop][locus][anc_allele] and min_gen >= introgression_generation:\n",
    "#                         idx = x_freqs[other_pop][locus][anc_allele].index(introgression_generation)\n",
    "#                         print('<<<', idx)\n",
    "#                         spliced_x_freqs = x_freqs[other_pop][locus][anc_allele][:introgression_generation] + x_freqs[pop][locus][anc_allele]\n",
    "#                         spliced_y_freqs = y_freqs[other_pop][locus][anc_allele][:introgression_generation] + y_freqs[pop][locus][anc_allele]\n",
    "#                         freqs = dict(zip(spliced_x_freqs, spliced_y_freqs))                        \n",
    "#                         print(\">\", spliced_x_freqs)\n",
    "#                     else:\n",
    "#                         freqs = dict(zip(x_freqs[pop][locus][anc_allele], y_freqs[pop][locus][anc_allele]))\n",
    "\n",
    "#                     pwdists[locus][a][b] = expected_coal_time(max_gen, min_gen, pop1_size_history, freqs)\n",
    "\n",
    "\n",
    "#     # print(pwdists[locus])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d2a910e-adbc-4474-96c7-0f862335c022",
   "metadata": {},
   "outputs": [],
   "source": [
    "#NODES[locus]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43e53db5-3a79-4525-adca-56c64d6e39dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(10, 3))\n",
    "plt.plot(sex_ratio_std['pop1'])\n",
    "plt.plot(sex_ratio_std['pop2'])\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6794a3c9-d36c-4326-a3f0-58ed1fc2e9b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(10, 3))\n",
    "plt.ylim(0, 1)\n",
    "plt.plot(male_ratio['pop1'])\n",
    "plt.plot(male_ratio['pop2'])\n",
    "plt.axhline(0.5)\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de3d71db-e63a-405a-84c4-d5fcf864e63b",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(11, 2, figsize=(15, 10), sharex=True, sharey=True)\n",
    "for i in range(10):\n",
    "    for p, pop in enumerate(['pop1', 'pop2']):\n",
    "        axes[i][p].plot(nr_x_alleles[pop][i])\n",
    "for p, pop in enumerate(['pop1', 'pop2']):\n",
    "    axes[-1][p].plot(nr_y_alleles[pop], color='red')\n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1997a2a-938b-4d21-a0b2-b94e01f8c2ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "fig, axes = plt.subplots(2, 5, figsize=(12, 5))\n",
    "for i, ax in enumerate(axes.flat):\n",
    "    c = ax.matshow(np.array(DRIVES[i])) ;\n",
    "    fig.colorbar(c, ax=ax, shrink=0.5)\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b04fa88-47ec-4f66-9df8-466ec08d76d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "pop = 'pop1'\n",
    "fig, axes = plt.subplots(10, 2, figsize=(15, 10), sharex=True, sharey=True)\n",
    "for ax, locus in zip(axes, range(LOCI)):\n",
    "    for i in x_freqs[pop][locus]:\n",
    "        if len(x_freqs[pop][locus][i]) < 50: # skip plotting alleles alive for less than 50 gens\n",
    "            continue\n",
    "\n",
    "        color = None\n",
    "        linewidth = None\n",
    "        if allele_val_maps[locus][i] == 2:\n",
    "            linewidth=3\n",
    "        else:\n",
    "            linewidth=1\n",
    "        if not i % 2:\n",
    "            ax[0].plot(x_freqs[pop][locus][i], y_freqs[pop][locus][i], linewidth=linewidth, color=color)\n",
    "            for g in introgression_generations:\n",
    "                ax[0].axvline(g)  \n",
    "        else:\n",
    "            ax[1].plot(x_freqs[pop][locus][i], y_freqs[pop][locus][i], linewidth=linewidth, color=color)\n",
    "            for g in introgression_generations:\n",
    "                ax[1].axvline(g)            \n",
    "sns.despine()  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf4c89e8-fc56-43d9-9251-7c8da515eaaa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "pop = 'pop2'\n",
    "fig, axes = plt.subplots(10, 2, figsize=(15, 10), sharex=True, sharey=True)\n",
    "for ax, locus in zip(axes, range(LOCI)):\n",
    "    for i in x_freqs[pop][locus]:\n",
    "        if len(x_freqs[pop][locus][i]) < 50: # skip plotting alleles alive for less than 50 gens\n",
    "            continue\n",
    "            \n",
    "        if allele_val_maps[locus][i] == 1:\n",
    "            linewidth=3\n",
    "        else:\n",
    "            linewidth=1            \n",
    "        if not i % 2:\n",
    "            ax[0].plot(x_freqs[pop][locus][i], y_freqs[pop][locus][i], linewidth=linewidth, color=color)\n",
    "            for g in introgression_generations:\n",
    "                ax[0].axvline(g)  \n",
    "        else:\n",
    "            ax[1].plot(x_freqs[pop][locus][i], y_freqs[pop][locus][i], linewidth=linewidth, color=color)\n",
    "            for g in introgression_generations:\n",
    "                ax[1].axvline(g) \n",
    "            \n",
    "sns.despine()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad3703cb-8eca-42f7-890e-b2778e32d2db",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# %matplotlib ipympl\n",
    "# fig, axes = plt.subplots(10, 2, figsize=(15, 10), sharex=True)\n",
    "# for ax, locus in zip(axes, range(LOCI)):\n",
    "#     for i in range(0, 50000, 2):\n",
    "#         if x_data[locus][i]:\n",
    "#             ax[0].plot(x_data[locus][i], y_data[locus][i])\n",
    "#     for i in range(1, 50000, 2):\n",
    "#         if x_data[locus][i]:\n",
    "#             ax[1].plot(x_data[locus][i], y_data[locus][i])\n",
    "# sns.despine()            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cee37570-6f6d-4386-8ec9-f0f434b318b6",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ## seed(7)\n",
    "\n",
    "# class Allele:\n",
    "#     # __slots__ = ('chrom', 'label', 'val')\n",
    "    \n",
    "#     def __init__(self, chrom, label, val):\n",
    "#         self.chrom = chrom\n",
    "#         self.label = label\n",
    "#         self.val = val\n",
    "    \n",
    "# class Male:\n",
    "#     # __slots__ = ('x_alleles', 'y_alleles')\n",
    "        \n",
    "#     def __init__(self, x_alleles, y_alleles):\n",
    "#         # assert x_allele.chrom == 'X'\n",
    "#         # assert y_allele.chrom == 'Y'\n",
    "#         self.x_alleles = x_alleles\n",
    "#         self.y_alleles = y_alleles\n",
    "\n",
    "#     def __add__(self, female):\n",
    "      \n",
    "#         # male gametes\n",
    "#         x_proportion = 0\n",
    "#         for locus in range(LOCI):\n",
    "#             x_drive = DRIVES[locus][self.x_alleles[locus].label][self.y_alleles[locus].label]\n",
    "#             y_drive = DRIVES[locus][self.y_alleles[locus].label][self.x_alleles[locus].label]\n",
    "#             x_proportion += x_drive / (x_drive + y_drive)\n",
    "#         x_proportion /= LOCI\n",
    "                            \n",
    "#         if random() < x_proportion:\n",
    "#             sperm_alleles = self.x_alleles[:]\n",
    "#         else:\n",
    "#             sperm_alleles = self.y_alleles[:]\n",
    "        \n",
    "#         # female gametes\n",
    "#         lst = []\n",
    "#         for a, b in zip(female.x1_alleles, female.x2_alleles):\n",
    "#             if random() < 0.5:\n",
    "#                 lst.append([a, b])\n",
    "#             else:\n",
    "#                 lst.append([b, a])\n",
    "#         x1_alleles, x2_alleles = map(list, zip(*lst))\n",
    "            \n",
    "#         if random() < 0.5:\n",
    "#             egg_alleles = x1_alleles\n",
    "#         else:        \n",
    "#             egg_alleles = x2_alleles\n",
    "  \n",
    "#         # make new individual\n",
    "#         if sperm_alleles[0].chrom == 'Y':\n",
    "#             return Male(egg_alleles, sperm_alleles)\n",
    "#         else:\n",
    "#             return Female(egg_alleles, sperm_alleles)\n",
    "            \n",
    "# class Female:\n",
    "#     # __slots__ = ('x1_alleles', 'x2_alleles')\n",
    "\n",
    "#     def __init__(self, x1_alleles, x2_alleles):\n",
    "#         # assert x1_allele.chrom == 'X'\n",
    "#         # assert x2_allele.chrom == 'X'\n",
    "#         self.x1_alleles = x1_alleles\n",
    "#         self.x2_alleles = x2_alleles\n",
    "\n",
    "#     def __add__(self, male):\n",
    "#         return male + self\n",
    "\n",
    "# LOCI = 10\n",
    "\n",
    "# DRIVES = [pd.DataFrame() for _ in range(LOCI)]\n",
    "\n",
    "# class Population:\n",
    "#     # __slots__ = ('males', 'females', 'cur_xlabel', 'cur_ylabel', 'alpha', 'beta', 'pop_size', 'drive_coef')\n",
    "            \n",
    "#     def __init__(self, males=[], females=[], cur_xlabel=[0 for _ in range(LOCI)], cur_ylabel=[1 for _ in range(LOCI)], alpha=5, beta=1):\n",
    "#         self.males = males\n",
    "#         self.females = females\n",
    "#         self.alpha = alpha\n",
    "#         self.beta = beta\n",
    "#         self.pop_size = len(self.males) + len(self.females)\n",
    "#         self.drive_coef = defaultdict(dict)\n",
    "#         self.cur_xlabel = cur_xlabel # label of latest x allele\n",
    "#         self.cur_ylabel = cur_ylabel # label of latest y allele\n",
    "        \n",
    "#     def initialize(self, pop_size):\n",
    "#         assert not pop_size % 2, 'initial pop_size must be even'\n",
    "#         self.males = [Male([Allele('X', self.cur_xlabel[locus], 1) for locus in range(LOCI)], \n",
    "#                            [Allele('Y', self.cur_ylabel[locus], 1) for locus in range(LOCI)]) for _ in range(pop_size//2)]\n",
    "#         self.females = [Female([Allele('X', self.cur_xlabel[locus], 1) for locus in range(LOCI)], \n",
    "#                                [Allele('X', self.cur_xlabel[locus], 1) for locus in range(LOCI)]) for _ in range(pop_size//2)]\n",
    "\n",
    "#         for locus in range(LOCI):\n",
    "#             self.update_drives(locus, self.cur_xlabel[locus])\n",
    "#             self.update_drives(locus, self.cur_ylabel[locus])\n",
    "\n",
    "# #     def update_drives(self, label):\n",
    "# #         # assert label >= len(self.DRIVES)\n",
    "\n",
    "# #         nrow = len(self.DRIVES[0]) if self.DRIVES else 1\n",
    "# #         row = np.random.normal(self.alpha, self.beta, size=nrow).tolist()\n",
    "# #         self.DRIVES.append(row)\n",
    "        \n",
    "# #         ncol = len(self.DRIVES)\n",
    "# #         col = np.random.normal(self.alpha, self.beta, size=ncol).tolist()\n",
    "# #         for row, val in zip(self.DRIVES, col):\n",
    "# #             row.append(val)\n",
    "            \n",
    "#     def update_drives(self, locus, label):  \n",
    "#         global DRIVES\n",
    "        \n",
    "#         assert label not in DRIVES[locus].columns.values, (label, DRIVES[locus].columns.values)\n",
    "#         nrows, ncols = DRIVES[locus].shape\n",
    "#         DRIVES[locus][label] = np.random.normal(self.alpha, self.beta, size=nrows)\n",
    "#         nrows, ncols = DRIVES[locus].shape\n",
    "#         new_row = pd.DataFrame.from_records([np.random.normal(self.alpha, self.beta, size=ncols)], \n",
    "#                                             columns=DRIVES[locus].columns)\n",
    "#         new_row.index = [label]\n",
    "#         DRIVES[locus] = pd.concat([DRIVES[locus], new_row])\n",
    "#         DRIVES[locus] = DRIVES[locus].copy() \n",
    "        \n",
    "#     def resample(self, new_pop_size=None):\n",
    "#         if new_pop_size is None:\n",
    "#             new_pop_size = self.pop_size\n",
    "#         new_males = list()\n",
    "#         new_females = list()\n",
    "        \n",
    "#         for i in range(new_pop_size):\n",
    "#             male = self.males[int(random() * len(self.males))]\n",
    "#             female = self.females[int(random() * len(self.females))]\n",
    "#             child = male + female\n",
    "                        \n",
    "#             if isinstance(child, Male):\n",
    "#                 new_males.append(child)\n",
    "#             else:\n",
    "#                 new_females.append(child)       \n",
    "                \n",
    "#         return Population(new_males, new_females, self.cur_xlabel, self.cur_ylabel)\n",
    "\n",
    "#     def mutate(self, x_mut_prob, y_mut_prob):\n",
    "\n",
    "#         for i in range(len(self.males)):\n",
    "#             for locus in range(LOCI):\n",
    "#                 if random() < x_mut_prob:\n",
    "#                     self.cur_xlabel[locus] += 2\n",
    "#                     self.males[i].x_alleles[locus] = Allele('X', self.cur_xlabel[locus], 1)\n",
    "#                     self.update_drives(locus, self.cur_xlabel[locus])\n",
    "\n",
    "#                 if random() < y_mut_prob:\n",
    "#                     self.cur_ylabel[locus] += 2\n",
    "#                     self.males[i].y_alleles[locus] = Allele('Y', self.cur_ylabel[locus], 1)\n",
    "#                     self.update_drives(locus, self.cur_ylabel[locus])\n",
    "\n",
    "#         for i in range(len(population.females)):\n",
    "#             for locus in range(LOCI):\n",
    "#                 if random() < x_mut_prob:\n",
    "#                     self.cur_xlabel[locus] += 2\n",
    "#                     self.females[i].x1_alleles[locus] = Allele('X', self.cur_xlabel[locus], 1)\n",
    "#                     self.update_drives(locus, self.cur_xlabel[locus])\n",
    "\n",
    "#                 if random() < x_mut_prob:\n",
    "#                     self.cur_xlabel[locus] += 2\n",
    "#                     self.females[i].x2_alleles[locus] = Allele('X', self.cur_xlabel[locus], 1)\n",
    "#                     self.update_drives(locus, self.cur_xlabel[locus])\n",
    "                        \n",
    "#     def x_freqs(self, locus):\n",
    "#         alleles = [male.x_alleles[locus].label for male in self.males] + \\\n",
    "#             [female.x1_alleles[locus].label for female in self.females] + \\\n",
    "#             [female.x2_alleles[locus].label for female in self.females]\n",
    "#         unique_elements, counts_elements = np.unique(np.array(alleles), return_counts=True)\n",
    "#         return unique_elements, counts_elements / counts_elements.sum()\n",
    "\n",
    "#     def y_freqs(self, locus):\n",
    "#         alleles = [male.y_alleles[locus].label for male in self.males]\n",
    "#         unique_elements, counts_elements = np.unique(np.array(alleles), return_counts=True)\n",
    "#         return unique_elements, counts_elements / counts_elements.sum()\n",
    "\n",
    "#     def male_freq(self):\n",
    "#         return len(self.males) / (len(self.males) + len(self.females))  \n",
    "    \n",
    "\n",
    "# # pop_size_history = [10000] * 10000\n",
    "# # x_mutation_rate = 0.0001\n",
    "# # y_mutation_rate = 0.0001\n",
    "# # generations = 5000\n",
    "\n",
    "# pop_size_history = [100] * 10000\n",
    "# x_mutation_rate = 0.01\n",
    "# y_mutation_rate = 0.01\n",
    "# generations = 500\n",
    "\n",
    "# # alpha = 1\n",
    "# # beta = 5\n",
    "# alpha = 5\n",
    "# beta = 1\n",
    "\n",
    "# _x = np.random.normal(alpha, beta, size=100000)\n",
    "# _y = np.random.normal(alpha, beta, size=100000)\n",
    "# null_dist = _x / (_x + _y)\n",
    "\n",
    "# population = Population(alpha=alpha, beta=beta)\n",
    "# population.initialize(pop_size_history.pop(0))\n",
    "\n",
    "# x_data = {}\n",
    "# y_data = {}\n",
    "# for locus in range(LOCI):\n",
    "#     x_data[locus] = collections.OrderedDict()\n",
    "#     y_data[locus] = collections.OrderedDict()\n",
    "#     nr_lines = 50000\n",
    "#     for i in range(nr_lines):\n",
    "#         x_data[locus][i] = []\n",
    "#         y_data[locus][i] = [] \n",
    "#     x_data[locus]['nr_x_alleles'] = []\n",
    "#     y_data[locus]['nr_x_alleles'] = []\n",
    "#     x_data[locus]['nr_y_alleles'] = []\n",
    "#     y_data[locus]['nr_y_alleles'] = []\n",
    "#     x_data[locus]['male_freq'] = []\n",
    "#     y_data[locus]['male_freq'] = []\n",
    "#     x_data[locus]['pop_size'] = []\n",
    "#     y_data[locus]['pop_size'] = []\n",
    "    \n",
    "# population_cache = {}\n",
    "    \n",
    "# for gen in trange(1, generations+1):\n",
    "    \n",
    "#     pop_size = pop_size_history.pop(0)\n",
    "#     population = population.resample(pop_size)\n",
    "#     population.mutate(x_mut_prob=x_mutation_rate, y_mut_prob=y_mutation_rate)\n",
    "#     if not gen % 10:\n",
    "#         population_cache[gen] = population\n",
    "    \n",
    "#     for locus in range(LOCI):\n",
    "#         x_alleles, x_freqs = population.x_freqs(locus)\n",
    "#         y_alleles, y_freqs = population.y_freqs(locus)\n",
    "\n",
    "#         if not gen % 100:\n",
    "#             live_labels = np.concatenate([x_alleles, y_alleles])\n",
    "#             DRIVES[locus] = DRIVES[locus].loc[live_labels, live_labels]     \n",
    "\n",
    "#         # plotting data\n",
    "#         for x_allele, x_freq in zip(x_alleles, x_freqs):\n",
    "#             x_data[locus][x_allele].append(gen)\n",
    "#             y_data[locus][x_allele].append(x_freq)\n",
    "#         for y_allele, y_freq in zip(y_alleles, y_freqs):\n",
    "#             x_data[locus][y_allele].append(gen)\n",
    "#             y_data[locus][y_allele].append(y_freq)\n",
    "#         # y_data[locus]['male_freq'].append(population.male_freq())\n",
    "#         # x_data[locus]['male_freq'].append(gen)\n",
    "#         # y_data[locus]['nr_x_alleles'].append(len(x_alleles))\n",
    "#         # x_data[locus]['nr_x_alleles'].append(gen)\n",
    "#         # y_data[locus]['nr_y_alleles'].append(len(y_alleles))\n",
    "#         # x_data[locus]['nr_y_alleles'].append(gen)\n",
    "#         # y_data[locus]['pop_size'].append(pop_size)\n",
    "#         # x_data[locus]['pop_size'].append(gen)  \n",
    "\n",
    "\n",
    "# import pickle\n",
    "\n",
    "# pickle.dump(population_cache, open( \"population_cache.p\", \"wb\" ))        \n",
    "        \n",
    "# %matplotlib inline\n",
    "# plt.matshow(np.array(DRIVES[0])) ;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a40144ed-6827-4d8d-b144-53f53eddfa8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fig, axes = plt.subplots(10, 2, figsize=(15, 10))\n",
    "# for ax, locus in zip(axes, range(LOCI)):\n",
    "#     for i in range(0, 50000, 2):\n",
    "#         if not x_data[locus][i]:\n",
    "#             break\n",
    "#         if len(x_data[locus][i]) > 5:\n",
    "#             ax[0].plot(x_data[locus][i], y_data[locus][i])\n",
    "#     for i in range(1, 50000, 2):\n",
    "#         if not x_data[locus][i]:\n",
    "#             break    \n",
    "#         if len(x_data[locus][i]) > 5:        \n",
    "#             ax[1].plot(x_data[locus][i], y_data[locus][i])\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3c078e0-bd69-4c66-bf0c-f73922300547",
   "metadata": {},
   "source": [
    "# if there is a lot of low freq alleles, is it then the high pi/higg ils regions that tend to sweep?\n",
    "\n",
    "safas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0130ddc4-b415-4686-8fd3-60d9f5cf1eba",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "\n",
    "frames = []\n",
    "for g in range(0, 1000, 10):\n",
    "    scatters = []\n",
    "    for i in range(1000):\n",
    "        scatters.append(go.Scatter(x=x_data[i][:g], y=y_data[i][:g]))\n",
    "    frames.append(go.Frame(data=scatters))\n",
    "        \n",
    "fig = go.Figure(\n",
    "    data=[go.Scatter(x=[0], y=[0]) for _ in range(1000)],\n",
    "    layout=go.Layout(\n",
    "        xaxis=dict(range=[0, 1000], autorange=False),\n",
    "        yaxis=dict(range=[0, 1], autorange=False),\n",
    "        updatemenus=[dict(\n",
    "            type=\"buttons\",\n",
    "            buttons=[dict(label=\"Play\",\n",
    "                          method=\"animate\",\n",
    "                          args=[None, {\"frame\": {\"duration\": 50, \n",
    "                                                 \"redraw\": False},\n",
    "                                                 \"fromcurrent\": True, \n",
    "                                                 \"transition\": {\"duration\": 10}}])])]\n",
    "\n",
    "                                            \n",
    "    ),\n",
    "    frames=frames\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c2ce089-701f-4e7b-b1fc-f24cab506635",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "\n",
    "xvals = np.linspace(1, 10, 100)\n",
    "yvals = np.sin(xvals)\n",
    "\n",
    "frames = []\n",
    "for i in range(len(xvals)):\n",
    "    frames.append(go.Frame(data=[go.Scatter(x=xvals[:i], y=yvals[:i]),\n",
    "                                 go.Scatter(x=xvals[:i], y=[p-0.2 for p in yvals[:i]])]))\n",
    "\n",
    "fig = go.Figure(\n",
    "    data=[go.Scatter(x=[0, 1], y=[0, 1]), go.Scatter(x=[0, 1], y=[0, 1])],\n",
    "    layout=go.Layout(\n",
    "        xaxis=dict(range=[0, 10], autorange=False),\n",
    "        yaxis=dict(range=[-1, 1], autorange=False),\n",
    "        updatemenus=[dict(\n",
    "            type=\"buttons\",\n",
    "            buttons=[dict(label=\"Play\",\n",
    "                          method=\"animate\",\n",
    "                          args=[None, {\"frame\": {\"duration\": 50, \n",
    "                                                 \"redraw\": False},\n",
    "                                                 \"fromcurrent\": True, \n",
    "                                                 \"transition\": {\"duration\": 0}}])])]\n",
    "\n",
    "                                            \n",
    "    ),\n",
    "    frames=frames\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "209c1897-c6a3-4d12-b98f-770356963166",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px\n",
    "\n",
    "df = px.data.gapminder()\n",
    "\n",
    "fig = px.bar(df, x=\"continent\", y=\"pop\", color=\"continent\",\n",
    "  animation_frame=\"year\", animation_group=\"country\", range_y=[0,4000000000])\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "474cb6cb-88cf-40bd-b31f-99cba94fe9ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "fig=go.Figure().set_subplots(2,1, vertical_spacing=0.05,\n",
    "                             specs=[[{\"type\": \"polar\"}], [{\"type\":\"bar\"}]])\n",
    "\n",
    "r=[3.5, 1.5, 2.5, 4.5, 4.5, 4, 3]\n",
    "theta=[65, 15, 210, 110, 312.5, 180, 270]\n",
    "width=[20,15,10,20,15,30,15]\n",
    "x= np.arange(1, 8)\n",
    "bh = [4, 6, 3, 7, 8, 5, 9]\n",
    "\n",
    "\n",
    "fig.add_trace(go.Barpolar(r=r[:1], theta=theta[:1], width=width[:1], \n",
    "                          marker_color=\"#ff8c00\",\n",
    "                          marker_line_color=\"black\",\n",
    "                          marker_line_width=1),1, 1)\n",
    "fig.add_trace(go.Bar(x=x[:1], y=bh[:1], marker_color=\"green\", width=0.95), 2,1);\n",
    "\n",
    "fig.update_layout(width=800, height=700, xaxis_range=[0.5, 7.5], yaxis_range=[0, max(bh)+0.5]);\n",
    "\n",
    "frames=[go.Frame(data=[go.Barpolar(r=r[:k+1], theta=theta[:k+1], width=width[:k+1]),\n",
    "                       go.Bar(x=x[:k+1], y=bh[:k+1])],\n",
    "                 name=f\"fr{k}\",\n",
    "                 traces=[0,1]) for k in range(len(r))]  #r, theta, width, x, bh must have the same len=number of frames\n",
    "\n",
    "fig.update(frames=frames)\n",
    "\n",
    "\n",
    "def frame_args(duration):\n",
    "    return {\n",
    "            \"frame\": {\"duration\": duration},\n",
    "            \"mode\": \"immediate\",\n",
    "            \"fromcurrent\": True,\n",
    "            \"transition\": {\"duration\": duration, \"easing\": \"linear\"},\n",
    "        }\n",
    "\n",
    "fr_duration=200  # customize this frame duration according to your data!!!!!\n",
    "sliders = [\n",
    "            {\n",
    "                \"pad\": {\"b\": 10, \"t\": 50},\n",
    "                \"len\": 0.9,\n",
    "                \"x\": 0.1,\n",
    "                \"y\": 0,\n",
    "                \"steps\": [\n",
    "                    {\n",
    "                        \"args\": [[f.name], frame_args(fr_duration)],\n",
    "                        \"label\": f\"fr{k+1}\",\n",
    "                        \"method\": \"animate\",\n",
    "                    }\n",
    "                    for k, f in enumerate(fig.frames)\n",
    "                ],\n",
    "            }\n",
    "        ]\n",
    "\n",
    "\n",
    "fig.update_layout(sliders=sliders,\n",
    "                  updatemenus = [\n",
    "                        {\n",
    "                        \"buttons\": [\n",
    "                            {\n",
    "                             \"args\": [None, frame_args(fr_duration)],\n",
    "                             \"label\": \"&#9654;\", # play symbol\n",
    "                             \"method\": \"animate\",\n",
    "                            },\n",
    "                            {\n",
    "                             \"args\": [[None], frame_args(fr_duration)],\n",
    "                             \"label\": \"&#9724;\", # pause symbol\n",
    "                             \"method\": \"animate\",\n",
    "                            }],\n",
    "                        \"direction\": \"left\",\n",
    "                        \"pad\": {\"r\": 10, \"t\": 70},\n",
    "                        \"type\": \"buttons\",\n",
    "                        \"x\": 0.1,\n",
    "                        \"y\": 0,\n",
    "                        }])\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "399577e6-2d8d-4ce2-b706-90619e3c2aa1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
